### ISSUE TITLE
Any suggestion about generating triplets efficiently?

### ISSUE THREAD

**Issue by alecwangcq** [Atcold/torch-TripletEmbedding#4]
Hi Canziani,
Thanks for your sharing the codes! It's very clear and neat. :)
But now I have no idea about generating triplets, especially when the dataset is very large (maybe Imagenet dataset).
Because there are 1000 categories and in total 1.2M images in Imagenet, so the way to generate triplets make a big difference to the convergence. 
Now I have a naive idea,  assume the batch size is 255.  First randomly sample 170 (2/3 \* 255) images from category K, and then equally divided them into two sets (set Anchor (85 images), and set Positive (85 images) ). Finally, sample 85 images from the remaining images, whose categories are not K, as set Negative.  
DO have any suggestion ?  💯

**Comment by bamos** [comment]
Hi @AlecWong - you might be interested in how we're selecting triplets in OpenFace and using them with this loss function. It works really well for my face dataset with ~500k images. The high-level ideas are in our [tech report](http://reports-archive.adm.cs.cmu.edu/anon/2016/CMU-CS-16-118.pdf) (and are similar to FaceNet's and VGG Face's ideas) and the training code is [here](https://github.com/cmusatyalab/openface/tree/master/training). Triplet selection is in [train.lua](https://github.com/cmusatyalab/openface/blob/master/training/train.lua).

**Comment by Atcold** [comment]
Alternatively, you could use a normal supervised training framework, remove the `nn.LogSoftMax()` and `nn.NLLCriterion()` and use the an online version of this loss function I've just refurbished [here](https://github.com/jhjin/triplet-criterion).
It expects a batch of embeddings, it performs a search for _anchor_, _positive_ and _negative_ samples, and `return`s a single gradient, as for the classical supervised training framework you're accustomed with.
